actors/src/main/scala/zio/actors/Actor.scala

```diff
@@ -2,11 +2,59 @@ package zio.actors
 
 import zio.actors.Actor.PendingMessage
 import zio.clock.Clock
+import zio.stream.ZStream
 import zio.{ Supervisor => _, _ }
 
+import scala.language.implicitConversions
+
 object Actor {
 
   private[actors] type PendingMessage[F[_], A] = (F[A], Promise[Throwable, A])
+  trait ActorResponse[R, S, +A] {
+    type ResponseType
+
+    def materialize[AA >: A](
+      state: Ref[S],
+      promise: Promise[Throwable, AA],
+      supervisor: Supervisor[R]
+    ): URIO[R with Clock, Unit]
+  }
+
+  object ActorResponse {
+    implicit def oneTime[R, S, A](response: RIO[R, (S, A)]): ActorResponse[R, S, A]                         =
+      new ActorResponse[R, S, A] {
+        override type ResponseType = RIO[R, (S, A)]
+
+        override def materialize[AA >: A](
+          state: Ref[S],
+          promise: Promise[Throwable, AA],
+          supervisor: Supervisor[R]
+        ): URIO[R with Clock, Unit] = {
+          val completer = ((s: S, a: A) => (state.set(s) *> promise.succeed(a)).ignore).tupled
+          response.foldM(
+            e =>
+              supervisor
+                .supervise[R, (S, A)](response, e)
+                .foldM(_ => promise.fail(e).ignore, completer),
+            completer
+          )
+        }
+      }
+    implicit def stream[R, S, E, I](response: ZStream[R, E, (S, I)]): ActorResponse[R, S, ZStream[R, E, I]] =
+      new ActorResponse[R, S, ZStream[R, E, I]] {
+        override type ResponseType = ZStream[R, E, (S, I)]
```

Comment:
I think that `ResponseType` here should be `ZStream[R, Throwable, (S, I)]`. 
As actors are location transparent and interaction can involve remote communication our interaction patterns `!` and `?` result in `Task[...]` (so a ZIO that can fail with `Throwable` as it might involve `write/readFromWire`) although in actor's behavior there could be no failures that can occur. 

Notice that here https://github.com/zio/zio-actors/pull/270/files#diff-c1a93d1001aa018540bc59b83dd7a09dR134 a stream is created from `Task[Any]` so resulting stream that is returned will be `Stream[Any, Throwable, Any]`. 

That is - a stream returned in our `RemoteSpec` will perform `readFromWire` to get elements (until receives `StreamEnd`) so every time it might fail with communication error but here it test the signature is `ZStream[Any, Nothing, String]` which seems that pulling elements out of stream can't fail.

I propose to remove `E` from `implicit def stream[R, S, E, I]` and fix returned `ZStream` signature to `ZStream[R, Throwable, I]`.

This might seem unclear but if we notice where communication occurs it should be clear - when we do `?` and the response it `oneTime` the call is synchronous and we wait for the response. 
When the response is a stream on the caller site we only create a recipe with `?` -> the communication occurs when we do `.runCollect` and this is a counterpart of getting response out of `?` with `oneTime`.

Please correct me if I got this wrong.

**EDIT**

### Problem

I think that support for streaming responses will require overhaul for remoting (due to the fact that we decide when we pull responses out of stream, so perform remote communication) as it breaks current design.

I introduced simple change to our test case:
```scala
_              <- actorRef ? Str("ZIO-Actor response... ")
resultStream   <- actorRef ? Strs(messages)
result         <- actorRef ? Str("ZIO-Actor response... ")
resultStream2  <- actorRef ? Strs(messages2)
```
And it fails - first we do the first ask -> remote actor system writes to the wire - we read it. 
Then we ask for a stream response and get the stream - the remote actor system writes all responses to the wire but we do not read them (!) but only define recipe for reading them. Then we ask for another oneTime `result` but we read a bit of those stream responses instead and it fails. This should definitely be covered in a test case like this (intertwine oneTime and stream interactions)

### Solution?

Off the top of my head - we can perform `runCollect` in `sendEnvelope` but it misses the point of #268 what you described as your goal.

You could just make it fully akka-like and pass `ActorRef` in a message you send (so you know where send messages back) - and perform all communication with `!` (tell) and drop stream responses.

WDYT?

Sorry for writing about this issue now - only now I had some time to think about it deeper.

Reply:
Regarding `ZStream` being a pull-based stream, but the responses being pushed over the wire:
Would it help to introduce a `Queue` instead of doing a `runCollect`?
That way the remote actor can push messages at any time, but from the client side we pull from the `Queue`.

I had also thought about the Akka way of passing an `ActorRef` and sending responses via `!`, but the resulting API would be quite a bit less elegant...

Reply:
Right now I think the issue on response producer site is:
```scala
(stream.map(StreamMsg) ++ ZStream(StreamEnd)).foreach(e =>
  writeToWire(worker, e)
)
```
If we immediately write stream elements to receiver like this then they need to be read right away (e.g. with runCollect), otherwise remoting can be easily broken (but it makes no sense as we are blocked until receive the whole stream).

I don't know a solution for this right now (maybe another abstraction layer for remote communication for atomic and stream responses where stream response elements will be stored on response producer site and only send over the wire when requested (but it would introduce other issues I think), or maybe something different)

Yeah, including sender ActorRef in protocol would make it less elegant but making stream responses that comply with current remoting feels to require major design change.

Reply:
Would it be possible to have a daemon on the receiver side that permanently tries to `readFromWire` and pushes into a `Queue` from which both `ZStream`s and answers from `?` can be read?
`StreamMsg` would then need some kind of "conversation identifier" so that when multiple streams are running at the same time their messages can be distinguished.

If we can come up with a working solution, would you see the overall goal as worthwhile or do you want to have a simple, no-bells-and-whistles library?

Reply:
Thanks, sounds great but I lack in knowledge and experience to elaborate on such solution (what are the pros and cons). Do you know a project where such solution is used?
As this is a design discussion also asking @mijicd @softinio. What do you think about it?